geomstats/geometry/siegel.py

```diff
@@ -0,0 +1,614 @@
+
+        ndim = len(point_shape)
+        if ndim == 3:
+            n_samples = point_shape[0]
+
+        point_transconj = ComplexMatrices.transconjugate(point)
+        aux = gs.einsum("...ij,...jk->...ik", point, point_transconj)
+
+        if ndim == 2:
+            eigenvalues = gs.linalg.eigvalsh(aux)
+            belongs = gs.all(eigenvalues <= 1 - atol)
+        elif ndim == 3:
+            belongs = []
+            for i_sample in range(n_samples):
+                eigenvalues = gs.linalg.eigvalsh(aux[i_sample, ...])
+                belongs.append(gs.all(eigenvalues <= 1 - atol))
+            belongs = gs.array(belongs)
+
+        if self.symmetric:
+            belongs = gs.logical_and(belongs, ComplexMatrices.is_symmetric(point))
+
+        return belongs
+
+    def projection(self, point, atol=gs.atol):
+        """Project a matrix to the Siegel space.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., n, n]
+            Matrix to project.
+        atol : float
+            Tolerance.
+            Optional, default: backend atol.
+
+        Returns
+        -------
+        projected: array-like, shape=[..., n, n]
+            Matrix in the Siegel space.
+        """
+        data_type = point.dtype
+        point_shape = gs.shape(point)
+
+        ndim = len(point_shape)
+        if ndim == 3:
+            n_samples = point_shape[0]
+
+        if self.symmetric:
+            point = ComplexMatrices.to_symmetric(point)
+
+        point_transconj = ComplexMatrices.transconjugate(point)
+        aux = gs.einsum("...ij,...jk->...ik", point, point_transconj)
+
+        projected = point
+        eigenvalues = gs.linalg.eigvalsh(aux)
+        max_eigenvalues = gs.max(eigenvalues, axis=-1) ** 0.5
+
+        if ndim == 2:
+            if max_eigenvalues >= 1 - atol:
+                projected *= gs.cast((1 - atol) / max_eigenvalues, dtype=data_type)
+
+        elif ndim == 3:
+            for i_sample in range(n_samples):
+                if max_eigenvalues[i_sample] > 1 - atol:
+                    projected *= gs.cast(
+                        (1 - atol) / max_eigenvalues[i_sample], dtype=data_type
+                    )
+
+        return projected
+
+    def random_point(self, n_samples=1, bound=1.0):
+        """Generate random points in the Siegel space.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+        bound : float
+            Bound of the interval in which to sample in the tangent space.
+            Optional, default: 1.
+
+        Returns
+        -------
+        samples : array-like, shape=[..., n, n]
+            Points sampled in the Siegel space.
+        """
+        n = self.n
+        size = (n_samples, n, n) if n_samples != 1 else (n, n)
+
+        samples = gs.random.rand(*size, dtype=gs.get_default_cdtype())
+        samples += 1j * gs.random.rand(*size, dtype=gs.get_default_cdtype())
```

Comment:
can now be simplified based in the last changes in the backend.

Reply:
Done.